project/Build.scala

```diff
@@ -675,6 +675,80 @@ object Build {
     recur(lines)
   }
 
+  val shadedSourceGenerator = (Compile / sourceGenerators) += Def.task {
+    val s = streams.value
+    val cacheDir = s.cacheDirectory
+    val dest = (Compile / sourceManaged).value / "downloaded"
+    val lm = dependencyResolution.value
+    
+    val dependencies = Seq(
+      ("com.lihaoyi", "pprint_3", "0.9.3"),
+      ("com.lihaoyi", "fansi_3", "0.5.1"),
+      ("com.lihaoyi", "sourcecode_3", "0.4.4"),
+    )
+
+    // Create a marker file that tracks the dependencies for cache invalidation
+    val markerFile = cacheDir / "shaded-sources-marker"
+    val markerContent = dependencies.map { case (org, name, version) => s"$org:$name:$version:sources" }.mkString("\n")
+    if (!markerFile.exists || IO.read(markerFile) != markerContent) {
+      IO.write(markerFile, markerContent)
+    }
+
+    FileFunction.cached(cacheDir / "fetchShadedSources",
+        FilesInfo.lastModified, FilesInfo.exists) { _ =>
+      s.log.info(s"Downloading and processing shaded sources to $dest...")
+      
+      if (dest.exists) {
+        IO.delete(dest)
+      }
+      IO.createDirectory(dest)
+
+      for((org, name, version) <- dependencies) {
+        import sbt.librarymanagement._
+
+        // Retrieve sources jar using dependencyResolution
+        val moduleId = ModuleID(org, name, version).sources()
+        val retrieveDir = cacheDir / "retrieved" / s"$org-$name-$version-sources"
+        
+        s.log.info(s"Retrieving $org:$name:$version:sources...")
+        val retrieved = lm.retrieve(moduleId, scalaModuleInfo = None, retrieveDir, s.log)
+        val jarFiles = retrieved.fold(
+          w => throw w.resolveException,
+          files => files.filter(_.getName.contains("-sources.jar"))
+        )
+
+        // Extract each retrieved jar
+        jarFiles.foreach { jarFile =>
+          s.log.info(s"Extracting ${jarFile.getName}...")
+          IO.unzip(jarFile, dest)
+        }
+      }
+
+      val scalaFiles = (dest ** "*.scala").get
+      scalaFiles.foreach { file =>
+        val text = IO.read(file)
+        if (!file.getName.equals("CollectionName.scala")) {
+          val processedText = "package dotty.shaded\n" +
+            text
```

Comment:
Could we have any kind of assert to make sure that patches are still applied? Something like a map with the replaced text as key and a counter maybe?